### PR TITLE
Add Source, a class for managing source files

### PR DESCRIPTION
**Pull request by zrlk**
This PR includes code to build a global mapping of
lightweight `Location` values to source text.

### CODE REVIEW

File: anodyne/base/source.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+#include "anodyne/base/source_buffer.h"
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace anodyne {
+
+/// \brief A point in a source file.
```
Comment (creachadair) [inline anodyne/base/source.h]: The name and the comment imply that there's some means of identifying the file—but really this is just an offset into some (notional) sequence of values, and the caller has to know which file it refers to via some external mechanism.

I don't think that requires any changes to the implementation, but the comment probably should avoid mentioning files.

File: anodyne/base/source.h [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+/// Locations should be passed by value.
+class Location {
+ public:
+  Location() : data_(0) {}
+  Location(const Location& o) : data_(o.data_) {}
+  Location& operator=(const Location& o) {
+    data_ = o.data_;
+    return *this;
+  }
+  bool is_valid() const { return data_ != 0; }
```
Comment (creachadair) [inline anodyne/base/source.h]: Please document the public members. In particular it'd be useful to mention what validity means.

Comment (creachadair) [reply anodyne/base/source.h]: And so on below, if you please…

File: anodyne/base/source.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  Location offset(int count) const {
+    Location o;
+    o.data_ = data_ + count;
+    return o;
+  }
+  uint32_t data() const { return data_; }
+
+ private:
+  friend class File;
```
Comment (creachadair) [inline anodyne/base/source.h]: ehhhhh…is there some way we can provision the Location with an API to make this friendship unnecessary?

File: anodyne/base/source.h (hunk 1 above)
Comment (zrlk) [reply anodyne/base/source.h]: done

File: anodyne/base/source.h (hunk 3 above)
Comment (zrlk) [reply anodyne/base/source.h]: done

File: anodyne/base/source.h (hunk 2 above)
Comment (zrlk) [reply anodyne/base/source.h]: done

**Pull request merged**